src/hb-ot-cmap-table.hh

```diff
@@ -1468,7 +1503,155 @@ struct cmap
+                              c->plan->glyphs_requested->has (p.second);
+                     })
+        | hb_map (hb_first)
+        ;
+
+        if (!format12_iter.len ())
+        {
+          format12_tables.del (_.first);
+          continue;
+        }
+
+        hb_set_t format4_unicodes_set;
+        hb_map_t format4_cp_glyphid_map;
+        (this+format4_tbl->subtable).collect_mapping (&format4_unicodes_set, &format4_cp_glyphid_map);
+
+        auto format4_iter =
+        + hb_zip (format4_unicodes_set.iter(), format4_unicodes_set.iter() | hb_map(format4_cp_glyphid_map))
+        | hb_filter (c->plan->_glyphset, hb_second)
+        | hb_filter ([c] (const hb_pair_t<hb_codepoint_t, hb_codepoint_t>& p)
+                     {
+                       return c->plan->unicodes->has (p.first) ||
+                              c->plan->glyphs_requested->has (p.second);
+                     })
+        | hb_map (hb_first)
+        ;
+
+        if (format12_iter.len () != format4_iter.len ())
+        {
+          format12_tables.del (_.first);
+          continue;
+        }
+
+        bool is_equal = true;
+        for (auto pair : + hb_zip (format12_iter, format4_iter))
+        {
+          if (pair.first != pair.second)
+          {
+            is_equal = false;
+            break;
+          }
+        }
+
+        if (!is_equal)
+        {
+          format12_tables.del (_.first);
+          continue;
+        }
+      }
+      else
+      {
+        hb_set_t format12_unicodes;
+        (this+format12_tbl->subtable).collect_unicodes (&format12_unicodes);
+
+        auto format12_iter =
+        + it
+        | hb_filter (format12_unicodes, hb_first)
+        | hb_map (hb_first)
+        ;
+
+        if (! format12_iter.len ())
+        {
+          format12_tables.del (_.first);
+          continue;
+        }
+
+        hb_set_t format4_unicodes;
+        (this+format4_tbl->subtable).collect_unicodes (&format4_unicodes);
+        auto format4_iter =
+        + it
+        | hb_filter (format4_unicodes, hb_first)
+        | hb_map (hb_first)
+        ;
+
+        if (format12_iter.len () != format4_iter.len ())
+        {
+          format12_tables.del (_.first);
+          continue;
+        }
+
+        bool is_equal = true;
+        for (auto pair : + hb_zip (format12_iter, format4_iter))
+        {
+          if (pair.first != pair.second)
+          {
+            is_equal = false;
+            break;
+          }
+        }
+
+        if (!is_equal)
+        {
+          format12_tables.del (_.first);
+          continue;
+        }
+      }
+    }
+
+    auto encoding_record_iter =
+    + encodingrec_iter
+    | hb_filter ([&] (hb_pair_t<const EncodingRecord&, unsigned> _)
+                 {
+                   unsigned format = (this+_.first.subtable).u.format;
+                   if (format != 12) return true;
+                   if (format12_tables.has (_.second)) return false;
+                   return true;
```

Comment:
slightly simpler to write:

return format != 12 || !format12_tables.has (_.second);